src/VisualStudio/Core/Def/Implementation/ColorSchemes/ColorSchemeApplier.ClassificationVerifier.cs

```diff
@@ -0,0 +1,212 @@
+
+            private ImmutableArray<string> Classifications { get; }
+
+            public ClassificationVerifier(IThreadingContext threadingContext, IServiceProvider serviceProvider, ImmutableDictionary<SchemeName, ColorScheme> colorSchemes)
+                : base(threadingContext)
+            {
+                _serviceProvider = serviceProvider;
+
+                _colorSchemes = colorSchemes.ToImmutableDictionary(
+                    nameAndScheme => nameAndScheme.Key,
+                    nameAndScheme => nameAndScheme.Value.Themes.ToImmutableDictionary(
+                        theme => theme.Guid,
+                        theme => theme.Category.Colors
+                            .Where(color => color.Foreground.HasValue)
+                            .ToImmutableDictionary(color => color.Name, color => color.Foreground!.Value)));
+
+                // Gather all the classifications from the core and scheme dictionaries.
+                var coreClassifications = DarkThemeForeground.Keys.Concat(BlueLightThemeForeground.Keys).Distinct();
+                var colorSchemeClassifications = _colorSchemes.Values.SelectMany(scheme => scheme.Values.SelectMany(theme => theme.Keys)).Distinct();
+                Classifications = coreClassifications.Concat(colorSchemeClassifications).ToImmutableArray();
+            }
+
+            /// <summary>
+            /// Determines if any Classification foreground colors have been customized in Fonts and Colors.
+            /// </summary>
+            public bool AreClassificationsCustomized(SchemeName schemeName, Guid themeId)
+            {
+                AssertIsForeground();
+
+                // Ensure we are initialized
+                if (_fontAndColorStorage is null)
+                {
+                    _fontAndColorStorage = _serviceProvider.GetService<SVsFontAndColorStorage, IVsFontAndColorStorage>();
+                    _fontAndColorUtilities = (IVsFontAndColorUtilities)_fontAndColorStorage!;
+                }
+
+                // Make no changes when in high contast mode or in unknown theme.
+                if (SystemParameters.HighContrast ||
+                    !_colorSchemes.TryGetValue(schemeName, out var colorScheme) ||
+                    !colorScheme.ContainsKey(themeId))
+                {
+                    return false;
+                }
+
+                var coreThemeColors = (themeId == KnownColorThemes.Dark)
+                    ? DarkThemeForeground
+                    : BlueLightThemeForeground;
+
+                var colorSchemeTheme = colorScheme[themeId];
+
+                // Open Text Editor category for readonly access and do not load items if they are defaulted.
+                if (_fontAndColorStorage.OpenCategory(TextEditorMEFItemsColorCategory, (uint)__FCSTORAGEFLAGS.FCSF_READONLY) != VSConstants.S_OK)
+                {
+                    // We were unable to access color information.
+                    return false;
```

Comment:
Under what circumstances do we expect this to happen? Agree we should handle failure, but wonder if we should use NFE if this is bad but recoverable

Reply:
Since this is only used to display a warning in the Options dialog, I don't think a failure here is too bad.